### ISSUE TITLE
[RFC] Changing the scoping rules of rescue variables

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#12522]
The _exception variable_ in a `rescue` clause do not has its scope tied to the body of the rescue clause. In particular, this means that if a variable exists with the same name, that same variable gets assigned the exception. This example shows how it works. 

```crystal
e = 1
begin
  raise "pum"
rescue e
  puts e
end
puts e+1 # currently fails: e has type Exception | Int32
```

Ruby shares this behavior, the last line fails with `undefined method `+' for #<RuntimeError: pum> (NoMethodError)`.

To me, the `e` in `rescue e` has its scope limited to the body of the `rescue` clause, meaning that it should _shadow_ the `e` in the outer scope.

<details>
<summary>What it means for this code:</summary>

```crystal
e = 1
begin
  raise "pum"
rescue e
  puts e # here e is #<Exception:pum>
end
puts e+1 # OK, e is 1
```
</details>

In the unlikely scenario that an exception should escape the `rescue` block, then it should be explicitly assigned.

<details>
<summary>Exception breaking free</summary>

```crystal
e = 1
begin
  raise "pum"
rescue e2
  e = e2 # run free, e2!
end
typeof(e) # => Exception | Int32
```
</details>

BTW, I looked into it after noting what Python does (hint: it's _weird_).

<details>
<summary>Python code for the curious</summary>

```py
e = None
try:
	raise("pum")
except Exception as e:
	print(repr(e))
print(e) # NameError: name 'e' is not defined
```
Python _erases_ `e`!
</details>

**Comment by devnote-dev** [comment]
I don't believe that `rescue` should be allowed to accept existing variables to assign to, if you change your example code to the following then you get the appropriate compiler error:
```crystal
e : Int32
begin
  raise "pum"
rescue e # Error: type must be Int32, not Exception+
  puts e
end
puts e+1
```

**Comment by yxhuvud** [comment]
Hmm. A case not explicitly mentioned yet about yet is
```
begin
  raise "pum"
rescue e
  puts e
end
typeof(e) # => Exception | Nil

```

Here, I think it is weird that `e` escapes the exception block. However, contrast it to how it works with blocks:
```

x = 0

[1].each { x = 1 }

x # =>  1


[1].each { y = 1 }

y # Does not compile.

```

I think that is the behavior I'd prefer - that it doesn't escape by default but if the variable exist it will escape.

**Comment by straight-shoota** [comment]
@yxhuvud I think the difference to the first block example is that there is na explicit assignment to an existing variable. That should definitely assign to that variable.
But the rescue clause is not an assignment. I'd see it similar to a method parameter. It declares a name for the parameter, and optionally restricts its type.
